bccsp/pkcs11/pkcs11.go

```diff
@@ -72,6 +72,9 @@ func (csp *impl) getSession() (session pkcs11.SessionHandle, err error) {
 			logger.Warningf("Get session info failed [%s], closing existing session and getting a new session\n", err)
 			csp.ctx.CloseSession(session)
 			session, err = createSession(csp.ctx, csp.slot, csp.pin)
+
+			// since handles are only valid for the current session context, clear the p11Key cache
+			csp.p11Keys.clear()
```

Comment:
Pretty much everything else is a nit except for this.

I'm not sure that this is quite sufficient. The select is not atomic with respect to the processing in each of the blocks. The way this is structured, if we get a session and think it's stale, we go off and close, recreate, and clear. If  this function can be called concurrently (and I don't see why it can't be) it can also do the same thing.

Now the cache clear and update are atomic but, depending on ordering, I think there are windows where the calls are interleaved such that a handle may appear to be valid but point to the wrong object. Don't know for sure, though.

Reply:
@sykesm  This is a tricky one and was one reason I was considering basing cache invalidation on invalid object handle errors rather than wiping out the entire cache for an invalid session.

For all sessions created in the same login context, the object handle for any given key will be the same (as far as I can tell from the spec) and any call to find key will return the same object handle for any given key.

An assumption I've made here is that if a session is invalid and when createSession is called a new login context in created, then all other sessions created under the prior login context will be invalid as well.  And in this case, you should not (cannot) end up using an object handle created under one context under a different context.

An alternative here would be to have a per session cache.

Reply:
I think it's probably simpler than that. I need a small break from hitting my head on the table so I'll take a quick look.

Reply:
I suppose we could have the same mutex on impl and the cache?  Then we could have an interleaved lock across both objects?